source/language/types.rst

```diff
@@ -373,3 +375,126 @@ variables whose values may only be known at run time.
+
+Casting from bool
+~~~~~~~~~~~~~~~~~
+
+``bool`` values cast from ``false`` to ``0.0`` and from ``true`` to ``1.0`` or 
+an equivalent representation. ``bool`` values can only be cast to ``bit[1]`` 
+(a single bit), so explicit index syntax must be given if the target ``bit``
+has more than 1 bit of precision.
+
+Casting from int/uint
+~~~~~~~~~~~~~~~~~~~~~
+
+``int[n]`` and ``uint[n]`` values cast to the standard types mimicking C99
+behavior. Casting to ``bool`` values follows the convention ``val != 0``.
+As noted above, if the value is too large to be represented in the
+target type the result is implementation-specific. Casting to ``bit[m]`` 
+is only allowed when ``m==n``. If the target 
+``bit`` has more or less precision, then explicit slicing syntax must be given.
+As noted, the conversion is done assuming a little-endian 2's complement 
+representation.
+
+Casting from float
+~~~~~~~~~~~~~~~~~~
+
+``float[n]`` values cast to the standard types mimicking C99 behavior (*e.g.* 
+discarding the fractional part for integer-type targets). As noted above, 
+if the value is too large to be represented in the 
+target type the result is implementation-specific. 
+Casting a ``float`` value to an ``angle[m]`` is accomplished by first 
+performing a modulo 2π operation on the float value. The resulting value 
+is then converted to the nearest ``angle[m]`` possible. In the event of a 
+tie between two possible nearest values the result is the one with an even 
+least significant bit (*i.e.* round to nearest, ties to even).
+
+Casting from angle
+~~~~~~~~~~~~~~~~~~
+
+``angle[n]`` values cast to ``bool`` using the convention ``val != 0.0``. 
+Casting to ``bit[m]`` values is only allowed when 
+``n==m``, otherwise explicit slicing synatx must be provided.
+
+When casting between angles of differing precisions (``n!=m``): if the target 
+has more significant bits, then the value is padded with ``m-n`` least 
+significant bits of ``0``; if the target has fewer significant bits, then 
+there are two acceptable behaviors that can be supported by compilers: 
+rounding and truncation. For rounding the value is rounded to the nearest 
+value, with ties going to the value with the even least significant bit.
+Trunction is likely to have more hardware support. This behavior can be
+controlled by the use of a ``#pragma``.
+
+Casting from bit
+~~~~~~~~~~~~~~~~
+
+``bit[n]`` values cast to ``bool`` using the convention ``val != 0``. Casting to 
+``int[m]`` or ``uint[m]`` is done assuming a little endian 2's complement 
+representation, and is only allowed when ``n==m``, otherwise explicit slicing 
+syntax must be given. Likewise, ``bit[n]`` can only be cast to ``angle[m]`` 
+when ``n==m``, in which case an exact per-bit copy is done using little-endian 
+bit order. Finally, casting between bits of differing precisions is not 
+allowed, explicit slicing syntax must be given.
```

Comment:
We don't specify the size of the smallest addressable unit here, so when we talk about conversions assuming a little-endian ordering, I think we're opening ourselves up to a possibility that `int[n](bit_string)` produces different values on different hardware.  For example, if I have the bitstring `x = "01_10"`, then on a "2-bit address" machine the conversion `uint[4](x)` produces the value 9, but a "4-bit address" machine would read it as 6.

I'm guessing we wrote it like this with the idea that the bitstrings could be directly re-interpreted, but that seems unusual to me; in general high-level programming, I expect the bitstring `0b1_0000_0010` to be the number 258, but by these conventions, on a machine that addresses things in 8-bit bytes, I'm not even sure what it would become.  I guess it would actually interpret my 4-bit separators differently, and read it as `0b1000_0001_0[000_000]`, which it then converts to an unsigned as 129?

Maybe I'm completely confused here - it seems like I might be making a very simple mistake.

Reply:
I think for little endian ordering the entire string has to be interpreted as little endian, not just addressable unit by addressable unit. It is the responsibility of the compiler to produce the correct behavior on any given piece of hardware, whether that is an illusion supported by software or not.

Reply:
Michael and I talked privately, and the upshot is that I had the terminology backwards in my head when talking about bitstring literals.  So it was a very simple mistake.